pkg/deviceplugin/lister.go

```diff
@@ -66,17 +68,86 @@ func (ml macvtapLister) Discover(pluginListCh chan dpm.PluginNameList) {
 		plugins = append(plugins, macvtapConfig.Name)
 	}
 
-	pluginListCh <- plugins
+	if len(plugins) > 0 {
+		pluginListCh <- plugins
+	}
+	return config, nil
+}
+
+func discoverByLinks(pluginListCh chan dpm.PluginNameList, nsPath string) {
```

Comment:
I have mixed feelings about self-healing of this function. Can we make it return an error in case something bad happens? It is way easier for admin to tell that something is wrong if the pod keeps restarting with "Error" state.

Reply:
I see your point. Some notes about it though:
* The only thing that can fail is the subscription. It can do so initially and overtime. Are you talking about an initial error or overtime? Distinguishing between the two will require specific code.
* Either the lister or any plugin can fail. Would we do this only if there is an error on the lister? What about if there is a similar error in any of the plugins?
* If we do this, any error in any lister/plugin might kill perfectly working functionality in other lister/plugins.
* DPM does not provide any interface to gracefully shutdown so we have limited control on which state we would leave things. We would have to panic, right?

Let me know what you think of these.

Reply:
> It can fail initially and overtime. 

With this I mean that all things that can fail there can fail initially and overtime. And failures are related to system state and not to initial input. So I don't know how daemons usually handle it but I thought it would be a fair compromise to attempt to self heal and don't handle those same situations different whether it is an initial failure or not.

Reply:
Also, ListerInterface Discover does not have an error return type, just remembered now. So really no clues by DPM on what to do on error in any case. Only options are to return an empty plugin list or panic.

Reply:
s/panic/os.Exit

Reply:
Calling os.Exit on initial Discover errors.